### ISSUE TITLE
Initial set up of FitTrans at ISIS

### ISSUE THREAD

**Issue by mguthriem** [mguthriem/PEARLsandpit#1]
I've run a quick check to confirm that the scripts in the repository are working. Before modifying code too much I think it  might be helpful to make sure the basics are working for at ISIS, and this might be done most easily just by running SNAP data. It also allows for some experience of how it runs while we work out the software kinks. 

Here are some set-up instructions. Just comment if/when these don't work...

1) download this file: [SNAPtestdata.tar.gz](https://github.com/mguthriem/PEARLsandpit/files/7529038/SNAPtestdata.tar.gz) and unzip it to a convenient place "MyFolder"
2) Fire up workbench and add the location of the repository to the python path by going to the IPython interpreter and typing: 
```
import sys
sys.path.append('/thelocation/ofyour/repository/')

```
3) Execute the set-up script in IPython:
```
%run \MyFolder\FitTrans_SetUp_GUI.py
```
4) You can now run the algorithm by just typing ``FitTrans`` in the workbench algorithm pane

**Comment by MrCrab-2019** [comment]
Thanks for this Malcolm - very clear. One issue I have hit immediately, is that the setup doesn't work due to LMFIT not being provided natively with Workbench. I found an old discussion you had on this with Daniel on the [Mantid Forum](https://forum.mantidproject.org/t/lmfit-installation/658). So I will try to install first using pip. However, perhaps we should look at implementing the Mantid least-squares minimisation code, I assume you must have looked into this already, so perhaps there is a reason not to?

**Comment by mguthriem** [comment]
You're right, I'd forgotten about that issue. And it took me quite some time to get them to install LMFIT here at SNS. I agree that using the mantid LS code might be the best way to get a sustainable solution. I'll need to look into how it works though to know if it's a big job or not.

**Comment by MrCrab-2019** [comment]
I used the Fit() function in Mantid as part of an algorithm I put together for high T measurements on PEARL (rewriting Transfit from genie to python). It seemed quite straight forward to implement in that case, but it is too soon to tell whether it can easily work within your code. So I will keep reading through it for now!

**Comment by mguthriem** [comment]
I had a quick look and I don't think it will be trivial. LMFIT takes the fitting parameters as a python dictionary, which is incredibly helpful as there are >100 variables in the model. It looks like mantid needs these to be individually specified, which would be tedious. Other requirements are the ability to specify limits and constraints and that the minimisation function must be able to handle a callback function to output the chi2 on each pass. 

Maybe worth putting in a ticket to get LMFIT installed while we chew on those issues...

**Comment by mguthriem** [comment]
Another thing we need is control over number of iterations. more than once, the refinement ended up in an infinite loop of cycles and the only way out was to kill mantid.

The documentation on fitting in mantid is threadbare. I'll ask the developers here if anyone knows much about it.

**Comment by mguthriem** [comment]
So, I got a reply from one of the developers that the easiest route will be to install lmfit. What platform do you use at ISIS? Still windows? The developer offered the following installation instructions (for other platforms)
```
In Ubuntu and installing for the native python, type in a terminal: sudo apt install python3-lmfit
In a virtual environment, type in a terminal:: pip install lmfit
In a conda environment. type in a terminal:: conda install lmfit
```

**Comment by MrCrab-2019** [comment]
Yep, I was able to get pip to install lmfit for now. I believe that the fitting routine can also call from a library of values, so should be possible to populate with a large number of peaks. Limits and constraints can also be accommodated. I think it may be possible to get it to work, but isn't something we need to worry about right now.

**Comment by mguthriem** [comment]
Great that you got it running! I did have some further developer input on using mantid. It seems that the only thing it can't do is the callback function. That's a bit of a bummer as it's satifying to watch the chi2 reduce in real time. They also mentioned that adding the library is possible but would require some work. I suggest we keep this on the table for now and decide later if it's necessary.

**Comment by mguthriem** [comment]
Another suggestion I got was to simply request that LMFIT be added to the mantid distribution. I can explore this too...
